### PR TITLE
TASK: Upgrade to Ember 2.2

### PR DESCRIPTION
**Pull request by GerDner**
./flow github:applyGerritChange 41296 + manually fix all merge conflicts

Discussion under https://discuss.neos.io/t/rfc-js-refactoring-process-of-backend/283

**Priority List Suggestion / Working Steps**
We propose to have intermediate steps which can be merged directly in Master; trying to avoid a long-running fork.

* migrate from require to Ember CLI (TODO WHO?)

* Update to Ember 2.0 based on the work done by @inkdpixels -- TODO: where is the repository which Tyl has worked on?

* start extracting the first simple components, test them (unit tests!) and use them throughout the code base.

* An example for simple components are button, dialog, ... (TODO WHO?)
(up to here, we have a pretty clear idea of what is happening. below, it is more unclear)

* learn from previous experiences, plan application structure (Model AND components)

* Start migrating application parts one-by-one.

**Original Commit Message:**
The following changes where made to full move from Ember 1.0 to Ember 2.0

 * Moved to the simpler bindAttr syntax
 * Moved to the simpler if syntax for bindAttr
 * Moved to the consistent handlebars action binding syntax
 * Moved all action methods into the new actions object in Ember.View's
 * Moved to the new handlebars each syntax
 * Fixed all elementId definitions in the Ember.View's
 * Moved from the old fallbackBinding= to the new fallback= property
 * Moved from the old idBinding= to the new id= property

### CODE REVIEW

**Comment by GerDner** [comment]
Not tested yet. 

First step after testing this revision would be to discuss the migration to ember-cli? 
e.g.
* app structure (using pods?)
* package-management (npm & bower)
* deprecation handling?

**Comment by GerDner** [comment]
First Checklist for the Ember Update.

- [x] Update Ember to 2.x (done, not pushed)
- [x] Update Ember-Template-Compiler to 2.x (done, not pushed)
- [x] Update Handlebars (done, not pushed)
- [ ] Update Gruntfile to handle the new framework versions correctly (the tricky part) - requirejs stuff
- [ ] Migrate all Ember.Views to Components (simple rename in the first step) since views are removed
- [ ] Migrate all ContainerView to Components implementing the Ember.MutableArray mixin
- [ ] Update use of {{controller}} since controller ist the default scope now and use of controller directly is removed
- [ ] Update use of @each in observers and computed propertiers (use [] instead)
- [ ] Refactor the Translate-Helper to conform HTMLBars

**Comment by GerDner** [comment]
Current Todos and known bugs:

- [ ] Fix editors like the image editor in the inspector (view creation and destruction) - very difficult!
- [ ] Fix Aloha Bugs 
- [ ] Fix Node constraints in the nodetree (empty strings in constraints)
- [ ] check select editors like the select2 wrapper - maybe check if refactoring is the better idea 
- [ ] fix upcoming deprecation warnings

**Comment by dimaip** [comment]
Hey guys, great work! How are things going lately with this?

One small note is that while fixing inspector editors it may make sense to think how they could be decoupled from the inspector itself and used in other parts of UI like [the node creation wizard](https://discuss.neos.io/t/rfc-allow-to-define-a-dialog-which-will-be-shown-on-node-creation/91/). It would require refactoring validation and some other stuff, so it's pretty complex, bad has to be done.

I'll be working on it again soon, so I wonder if I should rebase my work on top of this PR?

**Comment by GerDner** [comment]
We have big problems fixing the inspector and the image-editor. I think i would make sense to refactor the inspector with components.

**Comment by aertmann** [comment]
Hey everyone. Finally found time to focus on this one again, sorry for tha absence (had to focus on 2.1).

Rebased on master.

Basically container views were causing annoying problems and seems to be broken with the new eager view destruction. So replaced them completely, which makes sense since they will be removed anyway. This solves the property editors being rendered twice.

Additionally updated to the latest stable, since it works the same.

References inside templates behave differently now (especially after re-renders), so replaced some collections with each to avoid loosing those references.

Due to views being destroyed much more frequently, we need to handle asynchronous stuff carefully. E.g. image editor breaks during upload since the view is destroyed when changing the upload in progress variable, so we need to do some trickery to be able to update the value after the view has been destroyed since the logic resides inside the view. In general we need to completely remove such logic from the views, but for now I used workarounds to get it working as before. Lots of things are pretty difficult when not doing it the Ember way so when migrating to ember-cli, it should be much easier to refactor the components.

**Comment by aertmann** [comment]
There's still a lot of things that need to be adjusted, but there's nothing really broken anymore so each issue can be worked on separately and pushed independently. So feel free to pick one of the following known challenges and submit a fix :)

- [ ] migrate use of controller in actions targets e.g. {{action "someaction" target=controller}} -> component action handling
- [ ] Fix editors like the image editor in the inspector (view creation and destruction) - very difficult!
- [ ] Fix Aloha Bugs 
- [ ] Fix Node constraints in the node tree (empty strings in constraints)
- [ ] check select editors like the select2 wrapper - maybe check if refactoring is the better idea 
- [ ] fix upcoming deprecation warnings
- [ ] Inline handles cut/copy buttons missing classes
- [ ] workspace selector
- [ ] inspector duplicates editors
- [ ] plugin view editor
- [ ] asset editor
- [ ] select editor using ajax
- [ ] insert node panels (tree + inline)
- [ ] node info editor
- [ ] image / asset file upload drop zones

**Comment by dimaip** [comment]
Awesome! Will try to test it and rebase my wizard stuff on top of it as well.

Btw, maybe we could try to decouple editors from the inspector at the same go?

**Comment by aertmann** [comment]
@dimaip: To get this merged as quickly as possible, I'd avoid refactoring anything unless it's necessary to get it working with the new version. However it does make sense you base future development on top of this branch since it will avoid a lot of merge conflicts down the road.

Refactoring the editors might actually make the most sense doing at the same time as moving to ember-cli, since we'd have to find a new way to organize and load them anyway.

**Comment by skurfuerst** [comment]
@GerDner could you give me (skurfuerst) push access to your Development Collection so I can help out fixing the issues?

**Comment by GerDner** [comment]
you tried? because you already have full access.

**Comment by dimaip** [comment]
@aertmann agreed, will try to join in as time allows.

**Comment by skurfuerst** [comment]
@GerDner you are right, works :) I'll try to chime in in the upcoming week(s) to help out here!

**Comment by dimaip** [comment]
My first 2c, I think the node info editor now works...

**Comment by dimaip** [comment]
> @aertmann Fix editors like the image editor in the inspector (view creation and destruction) - very difficult!

What exactly isn't working with Image editor? To me everything seems fine.

**Comment by dimaip** [comment]
I've fixed the create node dialog, will push later this night

**Comment by dimaip** [comment]
I see that all of the actions on views are grouped in "actions" hash by @GerDner. From reading Ember docs, I believe it's only mandatory to do so for Controllers and not for views. When actions are put into "actions" hash on views, they loose reference to "this" of the view, and I don't know of an easy way to fix it.
I would leave actions on views as they were before, merge this the way they are, and then gradually migrate Views to Components.
Unless I really miss something obvious...

File: TYPO3.Neos/Resources/Public/JavaScript/Content/Components/TargetWorkspaceSelector.js [hunk 1]
```diff
@@ -22,20 +22,20 @@ define(
       classNameBindings: ['open:neos-open'],
       template: Ember.Handlebars.compile(template),
 
-      controller: TargetWorkspaceController,
-      nodeSelection: NodeSelection,
-      publishableNodes: PublishableNodes,
+			_controller: TargetWorkspaceController,
```
Comment (dimaip) [inline TYPO3.Neos/Resources/Public/JavaScript/Content/Components/TargetWorkspaceSelector.js]: @aertmann why do you rename controller here? I think this way it's no longer accessible in TargetWorkspaceSelector.html `content=targetWorkspaces`

**Comment by skurfuerst** [comment]
Hey @dimaip,

> When actions are put into "actions" hash on views, they loose reference to "this" of the view, and I don't know of an easy way to fix it.

I don't actually think so. "this" inside the actions hash is the same "this" as in other view methods -- at least that has worked for us in several big Ember projects.

All the best,
Sebastian

**Comment by dimaip** [comment]
@skurfuerst in controllers, not in views. Or maybe I'm just blind :)

File: TYPO3.Neos/Resources/Public/JavaScript/Content/Components/TargetWorkspaceSelector.js (hunk 1 above)
Comment (aertmann) [inline TYPO3.Neos/Resources/Public/JavaScript/Content/Components/TargetWorkspaceSelector.js]: had weird issues when using the ``controller`` variable, some magic got in the way.. can't remember exactly what though

**Comment by aertmann** [comment]
@dimaip: thanks for fixing those two :)

re image editor, had issues uploading images and cropping

re actions, AFAIK they have to be put inside actions to be called from other views since accessing methods and variables on views is now not really possible like it was earlier

**Comment by dimaip** [comment]
@skurfuerst 
image

I'm trying to fix AbstractModal.js. How do I access `destroy` view method from `cancel` action?

**Comment by dimaip** [comment]
@aertmann can I move it back to view in cases where we're calling that action from the same view?

**Comment by dimaip** [comment]
I guess the proper solution with actions would be to move to Components or at least to putting those actions to a controller (which I really don't like, but it seems to be the Ember way before Components).
We just have to get it working somehow and then move to Components. Components are cool, they remind me of React :)

**Comment by aertmann** [comment]
@dimaip it works fine for me, ``this`` is the view in that action.. however calling it using the esc binding needed to be adjusted.. see 507545186c3b38f00dad9f09374fd5f608631bd5

**Comment by dimaip** [comment]
@aertmann gosh, I'm an idiot... I tried to call it with `that.actions.cancel` and it had wrong `this` in that case.

**Comment by dimaip** [comment]
When document node is created, it does not go to edit title mode before creating.

**Comment by aertmann** [comment]
@dimaip: ah okay I see. your last comment is another issue right? could you add it to the checklist, so it's easier to keep an overview instead of having to read all comments?

**Comment by kitsunet** [comment]
Do we want to close this one?

**Comment by aertmann** [comment]
guess so, not much point merging finishing it, especially since it makes things worse than before until ember-cli is actually used..

**Pull request closed without merging**